### PR TITLE
Handle invalid HTML better

### LINKED ISSUE #2074: HTML email bodies not always displaying correctly
I identified a problem in that this new version is not fully accepting of 'all things HTML'.  My example is this:

### Steps to reproduce

An email body of the following body source:

```
<table border='0'> 
<tr><td><!==><!==>Hmailserver service shutdown:</td><td><!==><!==>Ok</td></tr> 
<tr><td><!==><!==>Spamassassin service shutdown:</td><td><!==><!==>Ok</td></tr> 
<tr><td><font color="red"><b>Spam Assassin Def Update:</td><td><font color="red"><b>Failed</td></tr> 
<tr><td><!==><!==>Spam Assassin Service startup: </td><td><!==><!==>Ok</td></tr> 
<tr><td><!==><!==>Hmailserver service startup:</td><td><!==><!==>Ok</td></tr> 
<tr><td><!==><!==>Zip:</td><td><!==><!==>Ok</td></tr> 
<tr><td><!==><!==>Email Cleardown Script: </td><td><!==><!==>Ok</td></tr> 
<tr><td><br></td></tr></table>See attached log file below for details. 
```

### Expected behavior

In the OLD version used to appear correctly as

> 
> Hmailserver service shutdown:	Ok
> Spamassassin service shutdown:	Ok
> **Spam Assassin Def Update:	Failed**
> Spam Assassin Service startup: 	Ok
> Hmailserver service startup:	Ok
> Zip:	Ok
> Email Cleardown Script: 	Ok
> 
> See attached log file below for details. 

(Where the bold is actually in red text)

### Actual behaviour

However, now in the new version it appeared as:

> 
> 
> **Spam Assassin Def Update:	Failed**
> 
> 
> 
> See attached log file below for details. 

Screenshot:

screenshot_2017-01-06-21-25-55

Note the missing lines (compared to the correct version).

I needed to change the source to replace the **<!==>** tags with  the correct REM tags

`<!--><!-->`

to make it all appear correctly.

ie,
`<tr><td>Hmailserver service shutdown:</td><td>Ok</td></tr>
.....`

Now some could argue that **<!==>** may not be official html tags and that causes the problem (which clearly they do in this new version).  My argument to this is that if they are unofficial HTML tags then they should be ignored and no attempt to interpret them should be made and therefore not alter the display in any way.....which is exactly what happened in previous version.

**So this version is not accepting of certain HTML formatting in some way - they have broken something and I suspect that it is this same change that is causing problems with HTML emails (not necessarily with them containing <!==> tags but who can know what else this version doesnt like rightly or wrongly)**

### Environment
K-9 Mail version:  v5.202

Android version:  4.2.1

Account type (IMAP, POP3, WebDAV/Exchange): IMAP

### PR DESCRIPTION
**Pull request by philipwhiuk**
Bump the version of HTMLCleaner (not technically required but there's been a few bug fixes) and add another setting - to decode `<!==>` into `&lt;!==&gt;`

Fixes #2074

### CODE REVIEW

File: k9mail/src/test/java/com/fsck/k9/helper/HtmlSanitizerTest.java [hunk 1]
```diff
@@ -99,4 +99,31 @@ public void shouldRemoveMetaRefreshButKeepOtherMetaTags() {
         assertEquals("<html><head><meta http-equiv=\"content-type\" content=\"text/html; charset=UTF-8\" /></head>" +
                 "<body>Message</body></html>", htmlSanitizer.sanitize(html));
     }
+
+    @Test
+    public void shouldProduceValidHtmlFromHtmlWithXmlDeclaration() {
+        String html = "<?xml version=\"1.0\" encoding=\"UTF-8\"?>" +
+                "<html><head></head><body></body></html>";
+        assertEquals("<html><head></head><body></body></html>", htmlSanitizer.sanitize(html));
```
Comment (cketti) [inline k9mail/src/test/java/com/fsck/k9/helper/HtmlSanitizerTest.java]: Please follow the arrange-act-assert structure. If you're concerned with consistency add another commit to rewrite the existing tests.

File: k9mail/src/test/java/com/fsck/k9/helper/HtmlSanitizerTest.java [hunk 2]
```diff
@@ -99,4 +99,31 @@ public void shouldRemoveMetaRefreshButKeepOtherMetaTags() {
...
+    }
+
+    @Test
+    public void shouldNormaliseTables() {
```
Comment (cketti) [inline k9mail/src/test/java/com/fsck/k9/helper/HtmlSanitizerTest.java]: Please use American English for consistency.

File: k9mail/src/test/java/com/fsck/k9/helper/HtmlSanitizerTest.java [hunk 3]
```diff
@@ -99,4 +99,31 @@ public void shouldRemoveMetaRefreshButKeepOtherMetaTags() {
...
+
+        assertEquals("<html><head></head><body><table><tbody>" +
+                "<tr><td></td><td></td></tr>" +
+                "</tbody></table></body></html>", htmlSanitizer.sanitize(html));
+    }
+
+    @Test
+    public void shouldHtmlEncodeXmlDirectives() {
+        String html = "<html><head></head><body><table>" +
+                "<tr><td>&lt;!==&gt;&lt;!==&gt;Hmailserver service shutdown:</td><td>&lt;!==&gt;&lt;!==&gt;Ok</td></tr>" +
```
Comment (cketti) [inline k9mail/src/test/java/com/fsck/k9/helper/HtmlSanitizerTest.java]: The input is already encoded. That's probably not what you wanted to test.

**Comment by philipwhiuk** [comment]
+1

File: k9mail/src/test/java/com/fsck/k9/helper/HtmlSanitizerTest.java [hunk 4]
```diff
@@ -96,7 +120,45 @@ public void shouldRemoveMetaRefreshButKeepOtherMetaTags() {
...
+    @Test
+    public void shouldHtmlEncodeXmlDirectives() {
+        String html = "<html><head></head><body><table>" +
+                "<tr><td><!==><!==>Hmailserver service shutdown:</td><td><!==><!==>Ok</td></tr>" +
+                "</table></body></html>";
+
+        String result = htmlSanitizer.sanitize(html);
+
+        assertEquals("<html><head></head><body><table><tbody>" +
+                "<tr><td>&lt;!==&gt;&lt;!==&gt;Hmailserver service shutdown:</td><td>&lt;!==&gt;<!==&gt;Ok</td></tr>" +
```
Comment (philipwhiuk) [inline k9mail/src/test/java/com/fsck/k9/helper/HtmlSanitizerTest.java]: Still wrong...

Comment (cketti) [inline k9mail/src/test/java/com/fsck/k9/helper/HtmlSanitizerTest.java]: Weird that Travis is not running the tests.

**Pull request merged**